File: 卡码/代码随想录/1/09-动态规划/Python/02-背包问题/01-01背包/Kama_0046.py
#46. 携带研究材料
from typing import *
import sys

def main():
  [m,n]=map(int,input().split())
  weight=list(map(int,input().split()))
  value=list(map(int,input().split()))
  dp:List[List[int]]=[[0]*(n+1) for x in range(m)]
  for j in range(weight[0],n+1):
    dp[0][j]=value[0]
  for i in range(1,m):
    for j in range(1,n+1):
      if j<weight[i]:
        dp[i][j]=dp[i-1][j]
      else:
        dp[i][j]=max(dp[i-1][j],dp[i-1][j-weight[i]]+value[i])
  print(dp[m-1][n])

if __name__=="__main__":
  main()

File: 卡码/代码随想录/JS/动态规划/02-背包问题/01-01背包/Kama_0046.js
//46. 携带研究材料
const rl=require("readline").createInterface({input:process.stdin});
const iter=rl[Symbol.asyncIterator]();
const readline=async ()=>(await iter.next()).value;

(async function(){
	const [m,n]=(await readline()).split(" ").map(Number);
	const weight=(await readline()).split(" ").map(Number);
	const value=(await readline()).split(" ").map(Number);
	const dp=new Array(m).fill(0).map(item=>new Array(n+1).fill(0));
	for(let j=weight[0];j<=n;j++){
		dp[0][j]=value[0];
	}
	for(let i=1;i<m;i++){
		for(let j=1;j<=n;j++){
			if(j<weight[i]){
				dp[i][j]=dp[i-1][j];
			}else{
				dp[i][j]=Math.max(dp[i-1][j],dp[i-1][j-weight[i]]+value[i]);
			}
		}
	}
	console.log(dp[m-1][n]);
})();